File: 20-actors/mitooshi/methods/test_ingest.py
#!/usr/bin/env python3
"""Tests for the mitooshi offline public-series normalizer (methods/ingest.py).

    PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python3 -m pytest test_ingest.py
    python3 test_ingest.py
"""
from __future__ import annotations

import json
import os
import pathlib
import subprocess
import sys

try:
    from ingest import normalize
except ImportError:
    from mitooshi.methods.ingest import normalize  # type: ignore

_HERE = pathlib.Path(__file__).resolve().parent
SAMPLE = _HERE.parent / "data" / "ingest" / "sample-batch.json"


def _batch():
    return json.loads(SAMPLE.read_text())


def test_sample_batch_normalizes_two_public_series():
    n = normalize(_batch())
    assert set(n["series"]) == {"s-hormuz-transit", "s-port-congestion"}


def test_proprietary_terminal_series_is_refused():
    n = normalize(_batch())
    assert len(n["refused"]) == 1
    r = n["refused"][0]
    assert r["id"] == "s-blocked-terminal" and "G4" in r["reason"]


def test_observations_are_sorted_append_only():
    n = normalize(_batch())
    hormuz = [o for o in n["obs"] if o[":obs/series"] == "s-hormuz-transit"]
    ats = [o[":obs/observed-at"] for o in hormuz]
    assert ats == sorted(ats)              # 非終末論: append-only, latest = current
    assert ats[-1] == 3 and hormuz[-1][":obs/value"] == 2.7


def test_categorical_class_preserved():
    n = normalize(_batch())
    cong = [o for o in n["obs"] if o[":obs/series"] == "s-port-congestion"]
    assert any(o.get(":obs/class") == "up" for o in cong)


def test_source_class_normalized_to_keyword():
    n = normalize(_batch())
    assert n["series"]["s-hormuz-transit"][":series/source-class"] == ":public-broadcast"
    assert n["series"]["s-hormuz-transit"][":series/sourcing"] == ":representative"


def test_live_flag_refused_without_operator_gate():
    env = dict(os.environ)
    env.pop("MITOOSHI_OPERATOR_GATE", None)
    p = subprocess.run([sys.executable, str(_HERE / "ingest.py"), "--live"],
                       capture_output=True, text=True, env=env)
    assert p.returncode != 0
    assert "G10" in (p.stdout + p.stderr) and "Council+operator gated" in (p.stdout + p.stderr)


def test_live_flag_gated_still_design_only_when_attested():
    env = dict(os.environ)
    env["MITOOSHI_OPERATOR_GATE"] = "1"
    p = subprocess.run([sys.executable, str(_HERE / "ingest.py"), "--live"],
                       capture_output=True, text=True, env=env)
    # attested but R0 → exits with the design-only message, NOT a live fetch
    assert "not implemented (design-only)" in (p.stdout + p.stderr)


def _run():
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
    print(f"ingest.py: {len(fns)}/{len(fns)} tests passed")
    return True


if __name__ == "__main__":
    sys.exit(0 if _run() else 1)

File: 20-actors/mitooshi/methods/test_ingest.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/test_ingest.py.
(ns mitooshi.methods.test-ingest
  "Tests for the mitooshi 見通し offline public-series normalizer (methods/ingest.clj).

  Covers offline normalization (G4 source-class membrane, G11 :representative honesty,
  append-only sorted observations) AND the G10 outward gate (live fetch refused without
  MITOOSHI_OPERATOR_GATE=1).

  Run:  bb --classpath 20-actors 20-actors/mitooshi/methods/test_ingest.clj"
  (:require [mitooshi.methods.ingest :as ing]
            [cheshire.core :as json]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [clojure.test :refer [deftest is run-tests]]))

(def ^:private this-file *file*)
(defn- batch []
  (json/parse-string
   (slurp (-> this-file io/file .getAbsoluteFile .getParentFile .getParentFile
              (io/file "data" "ingest" "sample-batch.json")))))

(deftest sample-batch-normalizes-two-public-series
  (let [n (ing/normalize (batch))]
    (is (= (set (keys (get n "series"))) #{"s-hormuz-transit" "s-port-congestion"}))))

(deftest proprietary-terminal-series-is-refused
  (let [n (ing/normalize (batch))]
    (is (= (count (get n "refused")) 1))
    (let [r (first (get n "refused"))]
      (is (= (get r "id") "s-blocked-terminal"))
      (is (str/includes? (get r "reason") "G4")))))

(deftest observations-are-sorted-append-only
  (let [n      (ing/normalize (batch))
        hormuz (filter #(= (get % ":obs/series") "s-hormuz-transit") (get n "obs"))
        ats    (map #(get % ":obs/observed-at") hormuz)]
    ;; 非終末論: append-only, latest = current
    (is (= ats (sort ats)))
    (is (= (last ats) 3))
    (is (= (get (last hormuz) ":obs/value") 2.7))))

(deftest categorical-class-preserved
  (let [n    (ing/normalize (batch))
        cong (filter #(= (get % ":obs/series") "s-port-congestion") (get n "obs"))]
    (is (some #(= (get % ":obs/class") "up") cong))))

(deftest source-class-normalized-to-keyword
  (let [n (ing/normalize (batch))]
    (is (= (get-in n ["series" "s-hormuz-transit" ":series/source-class"]) ":public-broadcast"))
    (is (= (get-in n ["series" "s-hormuz-transit" ":series/sourcing"]) ":representative"))))

;; G10 live-gate tests — mirror the watari pattern: pass env-gate as an explicit arg to
;; live-refusal (System/getenv is side-effecting; the impl exposes live-refusal as a pure fn
;; so tests can exercise both gate states without mutating process env).
(deftest live-flag-refused-without-operator-gate
  (is (some? (ing/live-refusal ["--live"] nil))
      "--live refused without the operator gate")
  (is (str/includes? (ing/live-refusal ["--live"] nil) "G10"))
  (is (str/includes? (ing/live-refusal ["--live"] nil) "Council+operator gated")))

(deftest live-flag-gated-still-design-only-when-attested
  ;; attested (gate="1") but R0 → exits with design-only message, NOT a live fetch
  (is (some? (ing/live-refusal ["--live"] "1")))
  (is (str/includes? (ing/live-refusal ["--live"] "1") "not implemented (design-only)")))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'mitooshi.methods.test-ingest)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
